notebooks/01-tessellate-ipu-tile-api-basics.ipynb

```diff
@@ -0,0 +1,572 @@
+    "\n",
+    "**Note:** Cycle count is not available on the IPU model simulator. `ipu_cycle_count` will always return a zeroed tensor on the IPU model simulator."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "b3519b88",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Start cycle count: TileShardedArray(array=DeviceArray([[391787498,         0],\n",
+      "             [391787499,         0],\n",
+      "             [391787499,         0]], dtype=uint32), tiles=(0, 1, 2)) (3, 2)\n",
+      "End cycle count: TileShardedArray(array=DeviceArray([[391787852,         0],\n",
+      "             [391787853,         0],\n",
+      "             [391787853,         0]], dtype=uint32), tiles=(0, 1, 2)) (3, 2)\n"
+     ]
+    }
+   ],
+   "source": [
+    "from tessellate_ipu import ipu_cycle_count\n",
+    "\n",
+    "\n",
+    "@jax.jit\n",
+    "def broadcast_add(lhs, rhs):\n",
+    "    # Tiles to split the workload on.\n",
+    "    tiles = tuple(range(len(lhs)))\n",
+    "    # Shard lhs on tiles (0, 1, ..., N)\n",
+    "    lhs = tile_put_sharded(lhs, tiles)\n",
+    "    # Replicate rhs on tiles (0, 1, ..., N)\n",
+    "    rhs = tile_put_replicated(rhs, tiles)\n",
+    "\n",
+    "    # Cycle count once inputs are sharded.\n",
+    "    lhs, rhs, start = ipu_cycle_count(lhs, rhs)\n",
+    "    # Map Poplar optimized `add` vertex to the sharded data.\n",
+    "    out = tile_map(jax.lax.add_p, lhs, rhs)\n",
+    "    # Cycle count after output is computed.\n",
+    "    out, end = ipu_cycle_count(out)\n",
+    "    return out, start, end\n",
+    "\n",
+    "\n",
+    "_, start, end = broadcast_add(lhs_data, rhs_data)\n",
+    "\n",
+    "print(\"Start cycle count:\", start, start.shape)\n",
+    "print(\"End cycle count:\", end, end.shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9886bfcd",
+   "metadata": {},
+   "source": [
+    "The function `ipu_cycle_count` returns raw cycle counts on every IPU tile, directly measured by the hardware. Note that `ipu_cycle_count` takes input arguments and returns them unchanged in order to provide control flow information to XLA and Poplar compilers (it measures cycle counts after these tensors have been computed).\n",
+    "\n",
+    "Tessellate IPU provides the raw values returned by IPU C++ intrinsics (https://docs.graphcore.ai/projects/poplar-api/en/2.4.0/ipu_builtins.html#get-count-l-from-csr), and this is why `start` and `end` tensors have `uint32` data types. As shown below, the raw cycle count can be easily translated into time performance figures. Please note timing measured in this way will differ significantly from the simple Python benchmarking of the `broadcast_add` function, as the latter will also include any JAX overhead, as well as all communication between the host and the IPU and IPU tile exchanges!"
```

Comment:
Is there any reason why you are referring to the 2.4.0 version of the doc?

Reply:
Fixed to refer to 3.3

Reply:
Thanks :+1: